internal/scheduler/schedulercontexts.go

```diff
@@ -0,0 +1,364 @@
+		)
+		sctx.NumScheduledJobs++
+	}
+	if qctx := sctx.QueueSchedulingContexts[jctx.Job.GetQueue()]; qctx != nil {
+		qctx.AddJobSchedulingContext(jctx, isEvictedJob)
+	}
+}
+
+// ClearJobSpecs zeroes out job specs to reduce memory usage.
+func (sctx *SchedulingContext) ClearJobSpecs() {
+	sctx.mu.Lock()
+	defer sctx.mu.Unlock()
+	for _, qctx := range sctx.QueueSchedulingContexts {
+		qctx.ClearJobSpecs()
+	}
+}
+
+func (sctx *SchedulingContext) SuccessfulJobSchedulingContexts() []*JobSchedulingContext {
+	sctx.mu.Lock()
+	defer sctx.mu.Unlock()
+	jctxs := make([]*JobSchedulingContext, 0)
+	for _, qctx := range sctx.QueueSchedulingContexts {
+		for _, jctx := range qctx.SuccessfulJobSchedulingContexts {
+			jctxs = append(jctxs, jctx)
+		}
+	}
+	return jctxs
+}
+
+// QueueSchedulingContext captures the decisions made by the scheduler during one invocation
+// for a particular queue.
+type QueueSchedulingContext struct {
+	// Time at which this context was created.
+	Created time.Time
+	// Executor this job was attempted to be assigned to.
+	ExecutorId string
+	// Queue name.
+	Queue string
+	// These factors influence the fraction of resources assigned to each queue.
+	PriorityFactor float64
+	// Total resources assigned to the queue across all clusters.
+	// Including jobs scheduled during this invocation of the scheduler.
+	ResourcesByPriority schedulerobjects.QuantityByPriorityAndResourceType
+	// Resources assigned to this queue during this scheduling cycle.
+	ScheduledResourcesByPriority schedulerobjects.QuantityByPriorityAndResourceType
+	// Job scheduling contexts associated with successful scheduling attempts.
+	SuccessfulJobSchedulingContexts map[string]*JobSchedulingContext
+	// Job scheduling contexts associated with unsuccessful scheduling attempts.
+	UnsuccessfulJobSchedulingContexts map[string]*JobSchedulingContext
+	// Protects the above maps.
+	mu sync.Mutex
+}
+
+func NewQueueSchedulingContext(queue, executorId string, priorityFactor float64, initialResourcesByPriority schedulerobjects.QuantityByPriorityAndResourceType) *QueueSchedulingContext {
+	if initialResourcesByPriority == nil {
+		initialResourcesByPriority = make(schedulerobjects.QuantityByPriorityAndResourceType)
+	} else {
+		initialResourcesByPriority = initialResourcesByPriority.DeepCopy()
+	}
+	return &QueueSchedulingContext{
+		Created:                           time.Now(),
+		Queue:                             queue,
+		ExecutorId:                        executorId,
+		PriorityFactor:                    priorityFactor,
+		ResourcesByPriority:               initialResourcesByPriority,
+		ScheduledResourcesByPriority:      make(schedulerobjects.QuantityByPriorityAndResourceType),
+		SuccessfulJobSchedulingContexts:   make(map[string]*JobSchedulingContext),
+		UnsuccessfulJobSchedulingContexts: make(map[string]*JobSchedulingContext),
+	}
+}
+
+const maxPrintedJobIdsByReason = 1
+
+func (qctx *QueueSchedulingContext) String() string {
+	var sb strings.Builder
```

Comment:
does this need a lock?

Reply:
No, the contexts themselves shouldn't be mutated while this function is called.